user/utest/exceptions.c

```diff
@@ -5,46 +5,38 @@
 #include <sys/signal.h>
 #include <stdlib.h>
 
-static int spawn_process(void (*proc_handler)(void)) {
-  int pid;
-  switch ((pid = fork())) {
-    case -1:
-      exit(EXIT_FAILURE);
-    case 0:
-      proc_handler();
-      exit(EXIT_SUCCESS);
-    default:
-      return pid;
-  }
-}
-
-static int check_exception(void (*code)(void), int signum) {
-  spawn_process(code);
-  int status;
-  wait(&status);
-  assert(WIFSIGNALED(status));
-  assert(signum == WTERMSIG(status));
-  return 0;
-}
+#include <stdio.h>
 
-static inline void exec_cp0_instr(void) {
+int test_exc_cop_unusable(void) {
   int value;
   asm volatile("mfc0 %0, $12, 0" : "=r"(value));
+  return 0;
 }
 
-static inline void exec_reserved_instr(void) {
+int test_exc_reserved_instruction(void) {
   /* The choice of reserved opcode was done based on "Table A.2 MIPS32 Encoding
    * of the Opcode Field" from "MIPS® Architecture For Programmers Volume II-A:
    * The MIPS32® Instruction Set" */
   asm volatile(".long 0x00000039"); /* objdump fails to disassemble it */
+  return 0;
 }
 
-int test_exc_cop_unusable(void) {
-  return check_exception(exec_cp0_instr, SIGILL);
+int test_exc_fpe(void) {
+  int d = __INT_MAX__;
+  asm volatile("addi %0, %0, 1" : : "r"(d));
```

Comment:
Causing floating point exception using integer overflow is a little bit unusual, and would not work on most of architectures (but works on MIPS). Maybe floating point division by zero is less confusing?

Reply:
Current version is ok. Following document: [signal.h](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/signal.h.html) lists following subclasses of `SIGFPE`:

 * `FPE_INTDIV` Integer divide by zero.
 * `FPE_INTOVF` Integer overflow.
 * `FPE_FLTDIV` Floating-point divide by zero.
 * `FPE_FLTOVF` Floating-point overflow.
 * `FPE_FLTUND` Floating-point underflow.
 * `FPE_FLTRES` Floating-point inexact result.
 * `FPE_FLTINV` Invalid floating-point operation.
 * `FPE_FLTSUB` Subscript out of range.

We don't send such an information to user-space yet.